### PR TITLE
Require AAL2 for platform admin MFA

### PR DESCRIPTION
**Pull request by teixr12**
## Summary

- Require `public.verify_mfa()` to rely on the Supabase `aal` claim instead of accepting any `amr.method = "otp"` session as MFA.
- Add regression coverage for an email OTP / magic-link style `aal1` session for a platform admin with a verified TOTP factor.
- Preserve access for a real `aal2` session while making the platform-admin wrapper require `aal2` before checking the admin secret.

## Root Cause

`is_platform_admin()` delegates to `verify_mfa()` before allowing platform-admin-only actions such as impersonation, admin credits, admin stats, and replication status.

`verify_mfa()` correctly required `aal2` when the user had a verified MFA factor, but it also returned true when the JWT `amr` array contained `method = "otp"`. Supabase email OTP / magic-link authentication can produce an OTP first-factor session while still staying at `aal1`, so mailbox access could satisfy the platform-admin MFA gate without the admin's TOTP factor.

## Security Impact

For users listed in `admin_users`, platform-admin endpoints should require an MFA-verified `aal2` session. This change prevents email OTP / magic-link sessions and zero-factor `aal1` sessions from being treated as platform-admin MFA.

## Test plan

- `git diff --check`
- Regression coverage added in `tests/is-admin-functions.test.ts` and `supabase/tests/07_auth_functions.sql`.
- I could not run the DB-backed Vitest test locally because this fresh environment does not have project dependencies installed and does not have `bun`, `supabase`, or `psql` available. CI is expected to exercise the added regression tests.

## Screenshots

N/A. This is a backend SQL authorization change with regression test coverage and no frontend/CLI UI changes.

## Checklist

- [ ] My code follows the code style of this project and passes `bun run lint:backend && bun run lint`.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] My change has adequate E2E test coverage.
- [ ] I have tested my code manually, and I have provided steps how to reproduce my tests.

Refs #1667
/claim #1667

## Summary by CodeRabbit

**New Features**
* TOTP (Time-based One-Time Password) multi-factor authentication is now available for users to enroll and verify
* Platform admin accounts now require additional authentication verification before access

**Tests**
* Added comprehensive test coverage for multi-factor authentication workflows
* Updated admin authentication helpers and test suites to support new authentication requirements

[Review Change Stack](https://app.coderabbit.ai/change-stack/Cap-go/capgo/pull/2103)

### CODE REVIEW

**Review by ShravanthReddy — COMMENTED** [review]
I found one test-path issue in the new regression test: the direct insert into `auth.mfa_factors` should hit the existing `trg_enforce_email_otp_for_mfa` trigger on a reset test DB. That trigger calls `public.is_recent_email_otp_verified()`, which only returns true when `public.user_security.email_otp_verified_at` is within the last hour, and the seed does not create a `user_security` row for the admin user.\n\nSo this statement is likely to fail before the assertions with `email otp verification required for mfa enrollment`:\n\n```sql\nINSERT INTO auth.mfa_factors (...)\nVALUES (..., 'verified'::auth.factor_status, ...);\n```\n\nA minimal fix would be to upsert a recent `public.user_security` row for `legacyAdmin` inside the transaction before inserting the MFA factor, similar to the existing delete-user reauth test helper. I also checked migration ordering locally with `scripts/check-supabase-migration-order.sh`; the timestamp itself is fine.

**Comment by teixr12** [comment]
Fixed the test setup issue by seeding a recent public.user_security.email_otp_verified_at row for the platform admin before inserting the verified MFA factor, so the existing MFA enrollment trigger can pass on reset test DBs.

**Comment by ShravanthReddy** [comment]
Thanks, this addresses the test setup issue I flagged. The added `public.user_security` upsert should allow the existing MFA enrollment trigger to pass before inserting the verified factor.

File: supabase/migrations/20260512003000_require_aal2_for_platform_admin_mfa.sql [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+          CASE
+            WHEN count(id) > 0 THEN array['aal2']
+            ELSE array['aal1', 'aal2']
+          END AS aal
+        FROM auth.mfa_factors
+        WHERE (SELECT auth.uid()) = user_id AND status = 'verified'
+    )
+  );
+END;
+$$;
```
[the remark this answers is not in the document]
Comment (teixr12) [reply supabase/migrations/20260512003000_require_aal2_for_platform_admin_mfa.sql:25]: Runtime/performance note for reviewers:

- Execution model: `public.verify_mfa()` is `SECURITY DEFINER` and is already used by existing restrictive RLS policies for authenticated-role access on `apikeys`, `app_versions`, `apps`, `channel_devices`, `channels`, `org_users`, and `orgs`.
- Scope of this PR: it does not add new RLS call sites or a new table lookup. It removes the previous `jsonb_array_elements(auth.jwt()->'amr')` OTP fallback branch and keeps the existing `auth.mfa_factors` user/status check as the deciding path.
- Cardinality expectation: MFA factors are per-user auth records, normally 0-1 verified TOTP factor for a user and bounded by the number of enrolled factors for the current authenticated user, not by tenant data rows.
- Index evidence: this repo's `supabase/schemas/prod.sql` references `auth.mfa_factors` but does not dump Supabase-managed auth schema indexes, so I cannot truthfully confirm the exact `(user_id, status)` index from repository evidence alone.
- EXPLAIN evidence: I cannot produce `EXPLAIN (ANALYZE, BUFFERS)` locally because this fresh environment has no project dependencies, `bun`, `supabase`, `psql`, or live `POSTGRES_URL`. I do not want to invent timings. The query to run in a maintainer DB is:

```sql
EXPLAIN (ANALYZE, BUFFERS)
SELECT CASE WHEN count(id) > 0 THEN array['aal2'] ELSE array['aal1', 'aal2'] END AS aal
FROM auth.mfa_factors
WHERE user_id = '<sample-user-id>'::uuid
  AND status = 'verified';
```

Given the current diff removes JSONB AMR expansion and does not broaden RLS invocation, I expect this to be no worse than the existing function and likely slightly cheaper per call. If maintainers want repository-backed proof in the PR itself, I can add a follow-up once a DB-backed `EXPLAIN` result is available.

**Comment by mySebbe** [comment]
I think there is still a zero-factor platform-admin gap in this version. `verify_mfa()` returns true for `aal1` whenever the caller has no verified rows in `auth.mfa_factors`, so a user listed in the `admin_users` secret with no verified MFA factor can still pass `public.is_platform_admin()` under a plain `aal1` session.

The new regression test covers an admin who already has a verified TOTP factor, but it does not cover the no-factor platform-admin case. If the intended invariant is "platform admin access requires MFA", this should either require `aal2` on the platform-admin path regardless of factor count or add a stricter helper for `is_platform_admin()`, plus a regression test where the admin has zero verified factors.

If no-factor platform admins are intentionally grandfathered, I would make that explicit in the PR description and function comment, because the current `is_platform_admin` comment says it requires MFA.

**Review by zinc-builds — APPROVED** [review]
## 🔍 Security Review — Hermes Agent

**Verdict: ✅ Approved** — The fix correctly addresses the MFA bypass and has good test coverage.

### What the PR fixes
The original `verify_mfa()` had a flaw: it returned `true` for any session containing `amr.method = "otp"`, even `aal1` magic-link/email-OTP sessions. A platform admin with a verified TOTP factor could access admin endpoints using only mailbox access — bypassing the MFA requirement entirely.

The fix removes the problematic `OR` clause and relies exclusively on the Supabase `aal` claim, which is the authoritative source for authentication assurance level.

### ✅ Confirmed correct
- The `CASE` expression correctly returns `array["aal2"]` when the user has MFA enrolled, and `array["aal1", "aal2"]` for users without MFA — so non-MFA users arent locked out of non-admin functionality
- `search_path` is set to empty string — prevents function injection via schema search path poisoning
- Grant statements match the existing convention (anon, authenticated, service_role)
- Test covers both the bypass case (`aal1 + otp → false`) and the legitimate case (`aal2 + totp → true`)

### ⚠️ Suggestions (non-blocking)
1. **SQL comment**: The migration replaces a non-trivial function with no inline comment explaining *why*. Consider adding a brief comment noting that `amr.method = "otp"` was intentionally removed because it accepted `aal1` sessions:
   ```sql
   -- v2: removed amr.method=otp fallback — email OTP / magic-link
   -- sessions at aal1 must not satisfy the platform-admin MFA gate.
   ```
2. **Test cleanup**: The test inserts into `auth.mfa_factors` inside a transaction but the `user_security` upsert and `mfa_factors` insert dont appear to be rolled back if `expect()` fails mid-test. Consider wrapping with a cleanup block or verifying that `withTransaction` handles rollback on assertion failures.
3. **Backward compatibility**: The migration file creates at timestamp `20260510200000` — confirm this is after all other migrations that reference `verify_mfa()` to avoid dependency ordering issues in fresh installs.

Nice work catching this — the `amr.method` vs `aal` distinction is subtle and requires understanding Supabases JWT claim model.

**Comment by teixr12** [comment]
Good catch, thank you. Fixed in 1176f2a by keeping public.verify_mfa() compatible with ordinary non-MFA users, but making the no-argument public.is_platform_admin() wrapper require an aal2 session before checking the admin_users secret. I also added a regression test for the zero-factor platform-admin aal1 case and added an inline SQL comment explaining why amr.method = 'otp' is not accepted for this gate.

**Comment by Abergelnathan** [comment]
Review note: `getAal2AuthHeadersForCredentials()` deletes every row in `auth.mfa_factors` for the signed-in user before enrolling its test TOTP factor. Because the admin stats/credits suites reuse shared admin credentials and Vitest can run files concurrently, this helper can race with another suite that is using the same admin account or remove MFA fixtures that were not created by this helper. It also leaves the newly verified factor behind, so repeated runs can accumulate state unless the next call happens to delete it first.

Could we scope this cleanup to factors created by this helper, for example by using a deterministic `friendlyName` prefix and deleting only that prefix, or create a dedicated per-suite admin user/factor and clean it up in `afterAll`? That keeps the new AAL2 coverage while avoiding cross-test interference around a security-sensitive table.

**Review by mingisrookie — COMMENTED** [review]
One touched-scope lint blocker remains in the new TOTP helper. Running:

```text
bun x eslint tests/test-utils.ts tests/is-admin-functions.test.ts tests/admin-credits.test.ts tests/admin-stats.test.ts tests/audit-logs.test.ts
```

reports these PR-local errors from the added `decodeBase32()` / `generateTotp()` code:

```text
tests/test-utils.ts
  336:39  error  Unexpected use of the global variable 'Buffer'. Use 'require("buffer").Buffer' instead  node/prefer-global/buffer
  351:10  error  Unexpected use of the global variable 'Buffer'. Use 'require("buffer").Buffer' instead  node/prefer-global/buffer
  357:18  error  Unexpected use of the global variable 'Buffer'. Use 'require("buffer").Buffer' instead  node/prefer-global/buffer
```

The same command also reports two pre-existing style/export findings near the top of `tests/test-utils.ts`, but the `Buffer` findings are introduced by this PR's helper. A minimal fix should be to import `Buffer` from `node:buffer` and use that imported binding in the helper, then rerun the touched-file eslint command before merge.

**Review by mingisrookie — COMMENTED** [review]
Rechecked current head `14f4242`: the PR-local `node/prefer-global/buffer` lint issue I flagged is resolved by importing `Buffer` from `node:buffer` and using that binding in the TOTP helper.

Verification:
- `NODE_OPTIONS=--max-old-space-size=4096 bun x eslint tests/test-utils.ts tests/is-admin-functions.test.ts tests/admin-credits.test.ts tests/admin-stats.test.ts tests/audit-logs.test.ts` now reports only the two pre-existing `tests/test-utils.ts` findings near the top of the file (`style/indent` on the existing multiline env fallback and `import/no-mutable-exports` on the existing `POSTGRES_URL` export); the three Buffer errors are gone.
- `git diff --check b11b7be..HEAD` passes.

I also attempted the focused `bun x vitest run tests/is-admin-functions.test.ts`, but this local checkout does not have Supabase Postgres running (`ECONNREFUSED 127.0.0.1:54322`), so I cannot truthfully report DB-backed test results from here.

**Review by KCDaemon — APPROVED** [review]
Rechecked latest head (14f4242). The platform-admin wrapper now requires the JWT `aal` claim to be `aal2` before it checks the admin secret, while `verify_mfa()` stays compatible with ordinary zero-factor users.

The earlier shared-test-factor cleanup concern is also addressed: the helper only deletes stale factors with the Capgo test prefix and then removes the exact enrolled factor in `finally`. GitHub CI is green, and I do not see a remaining blocker.
